File: cpp/460_LFU_Cache.cpp
/**
 * Difficult. We use 3 maps and a valuable min to record the minimum frequency in the cache.
 * 1st map: key -> (value, frequency)
 * 2nd map: frequency -> [key1, key2, ...]
 * 3rd map: key -> iterator of key in 2nd map
 */ 
class LFUCache {
 public:
  LFUCache(int capacity) {
    capacity_ = capacity;
  }
    
  int get(int key) {
    if (key_pair_map_.find(key) == key_pair_map_.end()) {
      return -1;
    } else {
      ++key_pair_map_[key].second;
      freq_keys_map_[key_pair_map_[key].second - 1].erase(key_iter_map_[key]);
      freq_keys_map_[key_pair_map_[key].second].push_back(key);
      key_iter_map_[key] = --(freq_keys_map_[key_pair_map_[key].second].end());
      if (min_ == key_pair_map_[key].second - 1 && freq_keys_map_[min_].empty()) {
        ++min_;
      }
      return key_pair_map_[key].first;
    }
  }
    
  void put(int key, int value) {
    if (capacity_ <= 0) { return; }
    get(key);
    if (key_pair_map_.find(key) != key_pair_map_.end()) {
      key_pair_map_[key].first = value;
    } else {
      if (size_ < capacity_) {
        if (min_ > 1) { min_ = 1; }
        key_pair_map_[key] = {value, 1};
        freq_keys_map_[1].push_back(key);
        key_iter_map_[key] = --(freq_keys_map_[1].end());
        ++size_;
      } else {
        int key_removed = *(freq_keys_map_[min_].begin());
        key_pair_map_.erase(key_pair_map_.find(key_removed));
        freq_keys_map_[min_].erase(freq_keys_map_[min_].begin());
        key_iter_map_.erase(key_iter_map_.find(key_removed));
        
        key_pair_map_[key] = {value, 1};
        freq_keys_map_[1].push_back(key);
        key_iter_map_[key] = --freq_keys_map_[1].end();
        min_ = 1;
      }
    }    
  }
 private:
  int capacity_ = 0;
  int size_ = 0;
  int min_ = 1;
  unordered_map<int,pair<int,int>> key_pair_map_;
  unordered_map<int, list<int>> freq_keys_map_;
  unordered_map<int, list<int>::iterator> key_iter_map_;
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
/**
 * ["LFUCache","put","get","put","get","get"]
    [[1],[2,1],[2],[3,2],[2],[3]]
 */

File: java/460_LFU Cache.java
/**
 * 难题，太难了，主要是很多细节。
 * 
 */
class LFUCache {

    static class Node {
        private int key;
        private int val;
        private int freq;
        private Node prev;
        private Node next;
    }
    
    private Map<Integer, Node> nodeByKey = new HashMap<>();
    private Map<Integer, Node> nodeByFreq = new HashMap<>();
    private Node head = new Node();
    private Node tail = new Node();

    final int capacity;
    
    public LFUCache(int capacity) {
        this.capacity = capacity;
        head.freq = -1;
        tail.freq = -1;
        tail.prev = head;
        head.next = tail;
        
    }
    
    public int get(int key) {
        //System.out.println("get " + key);
        Node node = nodeByKey.get(key);        
        if (node == null) { return -1; }
        Node nextNode = node.next;        
        // remove from nodeByFreq
        if (nodeByFreq.get(node.freq) == node) {
            if (node.next.freq == node.freq) {
                nodeByFreq.put(node.freq, node.next);
            } else {
                nodeByFreq.remove(node.freq);
            }
        } else {
            nextNode = nodeByFreq.get(node.freq);
        }        
        if (nodeByFreq.containsKey(node.freq + 1)) {
            nextNode = nodeByFreq.get(node.freq + 1);
        }
        // remove node from list
        node.prev.next = node.next;
        node.next.prev = node.prev;
                
        ++node.freq;                
        
        // add new freq to nodeByFreq
        node.next = nextNode;        
        node.prev = nextNode.prev;        
        node.prev.next = node;
        node.next.prev = node;
        nodeByFreq.put(node.freq, node);        
        return node.val;
    }
    
    public void put(int key, int value) {
        //System.out.println("put " + key + " " + value);
        if (capacity == 0) { return; }
        nodeByKey.putIfAbsent(key, new Node());
        
        if (nodeByKey.size() > capacity) {
            Node toRemove = tail.prev;
            // remove from nodeByFreq
            if (nodeByFreq.get(toRemove.freq) == toRemove) {                
                nodeByFreq.remove(toRemove.freq);
            }
            // remove from list
            toRemove.prev.next = tail;
            tail.prev = toRemove.prev;
            // remove from nodeByKey
            nodeByKey.remove(toRemove.key);
        }

        Node node = nodeByKey.get(key);
        node.key = key;
        node.val = value;        
        // new node inserted
        if (node.freq == 0) {
            node.prev = tail.prev;
            node.next = tail;
            node.prev.next = node;
            node.next.prev = node;
            nodeByFreq.put(node.freq, node);
        }        
        get(key);

    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */